### PR TITLE
[MONITOR] Data Collection Rules 2019-11-01-preview - New Cmdlets

### PR DESCRIPTION
**Pull request by herreraj-ms**
## Description

Onboarding AMCS (Azure Monitor Control Service) resources "data collection rules" (DCR) & "data collection rule associations" (DCRA) cmdlets.

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/761
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Monitor/Monitor.Test/Monitor.Test.csproj [hunk 1]
```diff
@@ -11,7 +11,7 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.Azure.Management.Monitor" Version="0.25.2-preview" />
```
Comment (msJinLei) [inline src/Monitor/Monitor.Test/Monitor.Test.csproj:14]: Please also update the Monitor sdk version in other modules' test project. Or the their tests will fail, which block the PR.

**Review by msJinLei — CHANGES_REQUESTED** [review]
Have you finish the cmdlet review?

File: src/Monitor/Monitor.Test/Monitor.Test.csproj (hunk 1 above)
Comment (herreraj-ms) [reply src/Monitor/Monitor.Test/Monitor.Test.csproj:14]: Yes, the latest it's 0.25.3-preview

**Comment by herreraj-ms** [comment]
> Have you finish the cmdlet review?

Yes: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/761

> 
> Please also fix the follow help file issue
> 
> ```
>  Line |
>   1564 |      $markdownFiles | ForEach-Object {
>        |                       ~~~~~~~~~~~~~~~~
>        | Exception calling "NodeModelToMamlModel" with "1" argument(s):
>        | "/Users/runner/work/1/s/artifacts/Debug/Az.Monitor/help/New-AzDataCollectionRule.md:28:(425) '```yaml {   "properties": {     "dataSources": ...'  Expect Heading"
> ```

Done! I remove the yaml section from the Description header

**Comment by msJinLei** [comment]
@herreraj-ms 

issues in generating help
```
WARNING : error encountered: Schema exception. This can occur when there are multiple code blocks in one example. D:\a\1\s\artifacts\Debug\Az.Monitor\help\New-AzDataCollectionRule.md:58:(688) '```yaml [D:\a\1\s\build.proj]
  {
    "properties": {
      "dataSources": ...'
   Expect Heading
  ForEach-Object: C:\Users\VssAdministrator\Documents\PowerShell\Modules\platyPS\0.14.0\platyPS.psm1:1564
  Line |
  1564 |      $markdownFiles | ForEach-Object {
       |                       ~~~~~~~~~~~~~~~~
       | Exception calling "NodeModelToMamlModel" with "1" argument(s):
       | "D:\a\1\s\artifacts\Debug\Az.Monitor\help\New-AzDataCollectionRule.md:58:(688) '```yaml {  
       | "properties": {     "dataSources": ...'  Expect Heading"

```

File: src/Monitor/Monitor/DataCollectionRules/SetAzureRmDataCollectionRuleCommand.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        [ValidateNotNullOrEmpty]
+        public string RuleName { get; set; }
+
+        /// <summary>
+        /// Gets or sets the data collection rule file definition path
+        /// </summary>
+        [Parameter(ParameterSetName = ByFile, Mandatory = true, ValueFromPipelineByPropertyName = false, HelpMessage = "The JSON file path.")]
+        [ValidateNotNullOrEmpty]
+        public string RuleFile { get; set; }
+
```
Comment (msJinLei) [inline src/Monitor/Monitor/DataCollectionRules/SetAzureRmDataCollectionRuleCommand.cs:71]: Please also provide byRuleId parameter set

File: src/Monitor/Monitor/DataCollectionRules/GetAzureRmDataCollectionRuleCommand.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        [ValidateNotNullOrEmpty]
+        public string ResourceGroupName { get; set; }
+
+        /// <summary>
+        /// Gets or sets the resource name parameter.
+        /// </summary>
+        [Parameter(ParameterSetName = ByName, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "The resource name")]
+        [Alias("Name")]
+        [ValidateNotNullOrEmpty]
+        public string RuleName { get; set; }
```
Comment (msJinLei) [inline src/Monitor/Monitor/DataCollectionRules/GetAzureRmDataCollectionRuleCommand.cs:55]: Please also provide byRuleId parameter set

**Review by msJinLei — CHANGES_REQUESTED** [review]

**Comment by herreraj-ms** [comment]
> @herreraj-ms
> 
> issues in generating help
> 
> ```
> WARNING : error encountered: Schema exception. This can occur when there are multiple code blocks in one example. D:\a\1\s\artifacts\Debug\Az.Monitor\help\New-AzDataCollectionRule.md:58:(688) '```yaml [D:\a\1\s\build.proj]
>   {
>     "properties": {
>       "dataSources": ...'
>    Expect Heading
>   ForEach-Object: C:\Users\VssAdministrator\Documents\PowerShell\Modules\platyPS\0.14.0\platyPS.psm1:1564
>   Line |
>   1564 |      $markdownFiles | ForEach-Object {
>        |                       ~~~~~~~~~~~~~~~~
>        | Exception calling "NodeModelToMamlModel" with "1" argument(s):
>        | "D:\a\1\s\artifacts\Debug\Az.Monitor\help\New-AzDataCollectionRule.md:58:(688) '```yaml {  
>        | "properties": {     "dataSources": ...'  Expect Heading"
> ```

Please check, I move the "extra" yaml section in the examples to the notes section

File: src/Monitor/Monitor/help/New-AzDataCollectionRule.md [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+Id                : /subscriptions/{subId}/resourceGroups/testdcr/providers/Microsoft.Insights/dataCollectionRules/newDcrEx1
+Name              : newDcrEx1
+Type              : Microsoft.Insights/dataCollectionRules
+Location          : East US 2 EUAP
+Tags              : {[tag2, value2], [tag1, value1]}
+```
+
+This command creates a data collection rules for the current subscription. [Note #1](#note1) has the content of the Rule File.
+
+### <a id="example2" name="example2"></a>Example 2: Create data collection rule, JSON from PSDataCollectionRuleResource
```
Comment (msJinLei) [inline src/Monitor/Monitor/help/New-AzDataCollectionRule.md:65]: incorrect header

**Review by msJinLei — CHANGES_REQUESTED** [review]

**Comment by herreraj-ms** [comment]
@msJinLei Just fix the help docs, please review!

File: src/Monitor/Monitor/help/New-AzDataCollectionRule.md [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+### System.String
+
+## OUTPUTS
+
+### Microsoft.Azure.Commands.Insights.OutputClasses.PSDataCollectionRuleResource
+
+## NOTES
+
+### Note #1: Content of C:\samples\dcrEx1.json
+```yaml
```
Comment (isra-fel) [inline src/Monitor/Monitor/help/New-AzDataCollectionRule.md:238]:
```suggestion
```json
```

File: src/Monitor/Monitor/help/New-AzDataCollectionRule.md [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+
+### System.String
+
+## OUTPUTS
+
+### Microsoft.Azure.Commands.Insights.OutputClasses.PSDataCollectionRuleResource
+
+## NOTES
+
+### Note #1: Content of C:\samples\dcrEx1.json
```
Comment (isra-fel) [inline src/Monitor/Monitor/help/New-AzDataCollectionRule.md:237]: I think platyPS (the markdown lib we use to generate help) doesn't like subtitles under "##Notes". Let's try plain text instead
```suggestion
Note #1: Content of C:\samples\dcrEx1.json
```

**Review by isra-fel — CHANGES_REQUESTED** [review]

**Comment by herreraj-ms** [comment]
@isra-fel fixed! Please review!

**Comment by isra-fel** [comment]
Hi @herreraj-ms I pulled your branch and digged into the help generation error, it seemed we had no choice but to put the json in the examples. I pushed a new commit that should fix the issue. Please let me know if it looks good to you

**Comment by herreraj-ms** [comment]
> Hi @herreraj-ms I pulled your branch and digged into the help generation error, it seemed we had no choice but to put the json in the examples. I pushed a new commit that should fix the issue. Please let me know if it looks good to you

Looks good to me!
Thanks

**Comment by isra-fel** [comment]
The only issue left is `-Tags`  should be named `-Tag` -- PowerShell is strict about cmdlet/parameter naming -- they should all be in singular form.

**Comment by isra-fel** [comment]
Kindly ping @herreraj-ms

**Comment by herreraj-ms** [comment]
> The only issue left is `-Tags` should be named `-Tag` -- PowerShell is strict about cmdlet/parameter naming -- they should all be in singular form.
> My applogies, this should be checked during cmdlet design review :P

Sorry for the delay!
Tag parameter change done, please review!

**Review by isra-fel — APPROVED** [review]
LGTM! Thanks

**Pull request merged**